src/Illuminate/Database/Query/Builder.php

```diff
@@ -2036,15 +2030,52 @@ protected function enforceOrderBy()
      */
     public function pluck($column, $key = null)
     {
-        $results = $this->get(is_null($key) ? [$column] : [$column, $key]);
+        $queryResult = $this->onceWithColumns(
+            is_null($key) ? [$column] : [$column, $key],
+            function () {
+                return $this->processor->processSelect($this, $this->runSelect());
+            }
+        );
+
+        if (empty($queryResult)) {
+            return collect();
+        }
 
         // If the columns are qualified with a table or have an alias, we cannot use
         // those directly in the "pluck" operations since the results from the DB
         // are only keyed by the column itself. We'll strip the table out here.
-        return $results->pluck(
-            $this->stripTableForPluck($column),
-            $this->stripTableForPluck($key)
-        );
+        $column = $this->stripTableForPluck($column);
+        $key = $this->stripTableForPluck($key);
+
+        if (is_array($queryResult[0])) {
+            return $this->pluckFromArrayColumn($queryResult, $column, $key);
+        } else {
+            return $this->pluckFromObjectColumn($queryResult, $column, $key);
+        }
+    }
+
+    /**
+     * Execute the given callback while selecting the given columns.
+     *
+     * After running the callback, the columns are reset to the original value.
+     *
+     * @param  array     $columns
+     * @param  callable  $callback
+     * @return mixed
+     */
+    protected function onceWithColumns($columns, $callback)
+    {
+        $original = $this->columns;
+
+        if (is_null($original)) {
+            $this->columns = $columns;
+        }
+
+        $result = $callback();
+
+        $this->columns = $original;
```

Comment:
Shouldn't this be wrapped in a `try`/`finally` block? We want to make sure we always reset them.

Reply:
It wasn't before my refactoring, so I don't want to change this here...